Src/Particle/AMReX_DenseBins.H

```diff
@@ -146,13 +223,250 @@ public:
+#pragma omp parallel for
+#endif
+        for (int j = 0; j < nchunks; ++j) {
+            int istart = j*chunksize;
+            int istop = (j == nchunks-1) ? nitems : (j+1)*chunksize;
+            for (int i = istart; i < istop; ++i) {
+                m_bins[i] = f(v[i]);
+                ++counts[nbins*j+m_bins[i]];
+            }
+        }
+
+#ifdef AMREX_USE_OMP
+#pragma omp parallel for
+#endif
+        for (int i = 0; i < nbins; ++i) {
+            index_type total = 0;
+            for (int j = 0; j < nchunks; ++j) {
+                auto tmp = counts[nbins*j+i];
+                counts[nbins*j+i] = total;
+                total += tmp;
+            }
+            m_counts[i] = total;
+        }
+
+        // note - this part has to be serial
+        m_offsets[0] = 0;
+        for (int i = 0; i < nbins; ++i) {m_offsets[i+1] = m_offsets[i] + m_counts[i];}
+
+#ifdef AMREX_USE_OMP
+#pragma omp parallel for
+#endif
+        for (int i = 0; i < nbins; ++i) {
+            for (int j = 0; j < nchunks; ++j) {
+                counts[nbins*j+i] += m_offsets[i];
+            }
+        }
+
+#ifdef AMREX_USE_OMP
+#pragma omp parallel for
+#endif
+        for (int j = 0; j < nchunks; ++j) {
+            int istart = j*chunksize;
+            int istop = (j == nchunks-1) ? nitems : (j+1)*chunksize;
+            for (int i = istart; i < istop; ++i) {
+                auto bid = m_bins[i];
+                m_perm[counts[nbins*j+bid]++] = i;
+            }
+        }
+
+        The_Arena()->free(counts);
+    }
+
+    /**
+     * \brief Populate the bins with a set of items.
+     *
+     * The algorithm is similar to a counting sort. First, we count the number
+     * of items in each bin. Then, we perform a prefix sum on the resulting counts.
+     * Finally, the set of partial sums is incremented in parallel using atomicInc,
+     * which results in a permutation array that places the items in bin-sorted order.
+     *
+     * This version uses a 3D box to specificy the index space over which the bins
+     * are defined.
+     *
+     * This overload uses the "Serial" parallelization strategy. It always runs in
+     * serial on the host, regardless of whether AMREX_USE_GPU or AMREX_USE_OMP
+     * has been defined.
+     *
+     * \tparam N the 'size' type that can enumerate all the items
+     * \tparam F a function that maps items to IntVect bins
+     *
+     * \param nitems the number of items to put in the bins
+     * \param v pointer to the start of the items
+     * \param bx the Box that defines the space over which the bins will be defined
+     * \param f a function object that maps items to bins
+     */
+    template <typename N, typename F>
+    void build (BinPolicy::SerialBinPolicy, N nitems, T const* v, const Box& bx, F&& f)
+    {
+        const auto lo = lbound(bx);
+        const auto hi = ubound(bx);
+        build(BinPolicy::Serial, nitems, v, bx.numPts(),
+              [=] AMREX_GPU_DEVICE (const T& t) noexcept
```

Comment:
Why is it marked with `AMREX_GPU_DEVICE` for the serial case?

Reply:
Just come copy / paste gone wrong. Removed.